flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/optimization/GradientDescent.scala

```diff
@@ -88,217 +55,296 @@ class GradientDescent() extends IterativeSolver {
+    learningRate: Double)
+  : DataSet[WeightVector] = {
+
+    data.mapWithBcVariable(currentWeights){
+      (data, weightVector) => (lossFunction.gradient(data, weightVector), 1)
+    }.reduce{
+      (left, right) =>
+        val (leftGradVector, leftCount) = left
+        val (rightGradVector, rightCount) = right
+        // Add the left gradient to the right one
+        BLAS.axpy(1.0, leftGradVector.weights, rightGradVector.weights)
+        val gradients = WeightVector(
+          rightGradVector.weights, leftGradVector.intercept + rightGradVector.intercept)
 
-    var weightVector: WeightVector = null
+        (gradients , leftCount + rightCount)
+    }.mapWithBcVariableIteration(currentWeights){
+      (gradientCount, weightVector, iteration) => {
+        val (WeightVector(weights, intercept), count) = gradientCount
 
-    @throws(classOf[Exception])
-    override def open(configuration: Configuration): Unit = {
-      val list = this.getRuntimeContext.
-        getBroadcastVariable[WeightVector](WEIGHTVECTOR_BROADCAST)
+        BLAS.scal(1.0/count, weights)
 
-      weightVector = list.get(0)
-    }
+        val gradient = WeightVector(weights, intercept/count)
 
-    override def map(example: LabeledVector): (Double, Int) = {
-      val lossFunction = parameters(LossFunction)
-      val predictionFunction = parameters(PredictionFunction)
+        val effectiveLearningRate = learningRate/Math.sqrt(iteration)
 
-      val loss = lossFunction.lossValue(
-        example,
-        weightVector,
-        predictionFunction)
+        val newWeights = takeStep(
+          weightVector.weights,
+          gradient.weights,
+          regularizationConstant,
+          effectiveLearningRate)
 
-      (loss, 1)
+        WeightVector(
+          newWeights,
+          weightVector.intercept - effectiveLearningRate * gradient.intercept)
+      }
     }
   }
 
-/** Calculates the regularized loss value, given the loss and the current weight vector
-  *
-  * The weight vector is received as a broadcast variable.
-  */
-private class RegularizedLossCalculation extends RichMapFunction[(Double, Int), Double] {
-
-  var weightVector: WeightVector = null
-
-  @throws(classOf[Exception])
-  override def open(configuration: Configuration): Unit = {
-    val list = this.getRuntimeContext.
-      getBroadcastVariable[WeightVector](WEIGHTVECTOR_BROADCAST)
-
-    weightVector = list.get(0)
-  }
-
-  override def map(lossAndCount: (Double, Int)): Double = {
-    val (lossSum, count) = lossAndCount
-    val regType = parameters(RegularizationType)
-    val regParameter = parameters(RegularizationParameter)
-
-    val regularizedLoss = {
-      regType.regLoss(
-        lossSum/count,
-        weightVector.weights,
-        regParameter)
+  /** Calculates the new weights based on the gradient
+    *
+    * @param weightVector
+    * @param gradient
+    * @param regularizationConstant
+    * @param learningRate
+    * @return
+    */
+  def takeStep(
```

Comment:
It would move us towards a more Breeze-like design. All FirstOrderMinimizers use the same optimization steps as seen [here](https://github.com/scalanlp/breeze/blob/master/math/src/main/scala/breeze/optimize/FirstOrderMinimizer.scala#L115)

I would vote for having a similar interface eventually.